src/main/kotlin/org/codefreak/codefreak/service/file/FileSystemFileService.kt

```diff
@@ -0,0 +1,414 @@
+
+    Files.walkFileTree(collectionPath, object : SimpleFileVisitor<Path>() {
+      override fun preVisitDirectory(dir: Path?, attrs: BasicFileAttributes?): FileVisitResult {
+        dir?.let {
+          val path = collectionPath.relativize(dir).normalize()
+          val normalizedPath = TarUtil.normalizeDirectoryName(path.toString())
+          TarUtil.mkdir(normalizedPath, tarOutput)
+        }
+        return FileVisitResult.CONTINUE
+      }
+
+      override fun visitFile(file: Path?, attrs: BasicFileAttributes?): FileVisitResult {
+        file?.let {
+          val path = collectionPath.relativize(file).normalize()
+          val normalizedPath = TarUtil.normalizeFileName(path.toString())
+          TarUtil.writeFileWithContent(normalizedPath, Files.newInputStream(it), tarOutput)
+        }
+        return FileVisitResult.CONTINUE
+      }
+    })
+
+    tarOutput.close()
+
+    return ByteArrayInputStream(output.toByteArray())
+  }
+
+  override fun writeCollectionTar(collectionId: UUID): OutputStream {
+    return object : ByteArrayOutputStream() {
+      override fun close() {
+        // Overwrite the existing collection with the tar contents
+        deleteCollection(collectionId)
+
+        TarArchiveInputStream(ByteArrayInputStream(toByteArray())).use { input ->
+          input.entrySequence().forEach {
+            if (it.isFile) {
+              if (it.size == 0L) {
+                createFiles(collectionId, setOf(it.name))
+              } else {
+                writeFile(collectionId, it.name).use { outputStream ->
+                  IOUtils.copy(input, outputStream)
+                }
+              }
+            } else if (it.isDirectory) {
+              createDirectories(collectionId, setOf(it.name))
+            }
+          }
+        }
+      }
+    }
+  }
+
+  override fun collectionExists(collectionId: UUID): Boolean {
+    val collectionPath = getCollectionPath(collectionId)
+    return Files.exists(collectionPath)
+  }
+
+  override fun deleteCollection(collectionId: UUID) {
+    val collectionPath = getCollectionPath(collectionId)
+
+    if (Files.exists(collectionPath) && Files.isDirectory(collectionPath)) {
+      deleteDirectoryRecursively(collectionPath)
+    }
+  }
+
+  override fun walkFileTree(collectionId: UUID): Sequence<FileMetaData> {
+    val collectionPath = getCollectionPath(collectionId)
+    val fileTree = mutableListOf<FileMetaData>()
+
+    Files.walkFileTree(collectionPath, object : SimpleFileVisitor<Path>() {
+      override fun visitFile(file: Path?, attrs: BasicFileAttributes?): FileVisitResult {
+        file?.let { fileTree.add(filePathToFileMetaData(it, collectionPath)) }
+        return FileVisitResult.CONTINUE
+      }
+
+      override fun postVisitDirectory(dir: Path?, exc: IOException?): FileVisitResult {
+        dir?.let { fileTree.add(filePathToFileMetaData(it, collectionPath)) }
+        return FileVisitResult.CONTINUE
+      }
+    })
+
+    return fileTree.asSequence()
```

Comment:
By using `File.walk` this should generate a sequence that creates new objects lazy.